One pager writeups (#11045)
dotnet/msbuild: documentation/specs/proposed/decoupling-vs-sdk.md

## Goals and Motivation

We are aiming for:
 - Consistent end-user eperience for build in either DotNet CLI or Visual Studio.
 - Decoupling the .NET SDK experience from Visual Studio 
 - Decoupling the .NET SDK from VS.

The reason we are persuing this is for a better experience when using or writting Roslyn analyzers and MSBuild Tasks. Currently tooling authors need to target NetStandard2.0 for their projects to be recognized by VS. Another options is to multi-target but it takes a lot more effort and time spent on that. Another aspect is the user experience, if the Roslyn version for VS analyzers and generators doesn't match the one in Visual Studio, they do not work.

## Impact
There are a few area of impact:
 - .NET SDK style project builds will be more stable between VS and CLI builds, as the tooling will not be devided between different versions.

>  - Reduced cost of development for external and internal teams that contribute to Roslyn Analyzers, SourceBuild generators, or MSBuild Tasks.

Reviewer: ```suggestion
 - Reduced cost of development for external and internal teams that author Roslyn Analyzers, source generators, or MSBuild Tasks.
```

>  - End-user will not experience mismatch between analyzer versions, and confirmation that their .NET SDK style builds will behave the same way in VS and in the command line.

Reviewer: ```suggestion
 - End-user will not experience mismatch between analyzer versions, and gain higher confidence that their .NET SDK project builds will behave the same way in VS and in the command line.
```

## Stakeholders

> For the internal stakeholder, we have the teams that will continue the work to fully complete the VS and .NET SDK decoupling feature after our base work is done. There are two handovers in this project:

Reviewer: ```suggestion
Other teams will need to work to fully complete the VS and .NET SDK decoupling feature after our base work is done. There are two handovers in this project:
```

> 1. Enabling the MSBuild.exe execution state to be the same as DotNet command line invocation so the Roslyn team can enable the use of their core compiler in VS.

Reviewer: ```suggestion
1. After providing enough information to do so through MSBuild and the SDK, Roslyn will need to use it to invoke their .NET compiler in VS.
```

> 2. Tasks and other projects can be written in .NET Core and the .NET SDK projects will build successuflly in VS. This enables other teams to migrate their tasks to .NET Core instead of keeping them targeting .NET Framework.

Reviewer: ```suggestion
2. After MSBuild enables tasks to target .NET even for VS use, task-owning teams like the .NET SDK will need to migrate their targets to use .NET Core instead of keeping them targeting .NET Framework.
```

> The handovers should allow other teams to proceed with their work smoothly and no change in build behaviour should be present within MSBuild.

Reviewer: ```suggestion
The handovers should allow other teams to proceed with their work smoothly and no unexpected change in build behavior should be present within MSBuild.
```

## Risks
A few risks associated with this feature:

>  - Our work is early in the development effort. If this feature is discovered to have too large of an impact in experience of performance the work might be delayed or discarded.

Reviewer: ```suggestion
 - If .NET Core tasks is discovered to have too large of a performance impact (due to IPC overhead to a .NET process), core partner teams may choose to keep multitargeting their tasks for improved user perf.
```

>  - There might be a performance hit on VS once we start running tasks on .NET Core. It would depending on the amount of non-framwork tasks that the project will need to load when opening it in VS. The performance gain from pre-loading will not be available in this scenario.

Reviewer: How is this different from the previous bullet?

>  - There are no concrete deadlines for our part of the feature, but we are aiming for an early preview cycle, so we have a chance to measure the consequences and fix any issues that arise.

Reviewer: There is a concrete deadline: 17.14. Usually we support the .NET SDK in the previous VS (when targeting older .NET versions), so if we don't hit 17.14 we'll have to either change that policy at the SDK level or not adopt the new stuff in 10.0.

## Plan
 1. Ensure that MSBuild.exe provides the same execution state as the dotnet command line invocation.
    -  This is should take around 1 dev week to complete, and will be handed over to Roslyn team.
 2. Implement .NET Core task host, tasks can be executed on the .NET Core vresion instead of .NET framework.
    - This should take 1 to 2 dev months to complete, including extensive testing. This would be handed over to internal teams that have custom tasks so they can be updated and tested.

>  3. Load common targets from the .NET SDK and not from .NET NetFramework on VS. This work depends on other team's finilizing their part of the feature and might not be in scope for .NET 10.

Reviewer: I don't think there's any way this is in scope for 10 and would preemptively cut it.

Author: It is already there that it is no in scope for 10. But I think we should leave it as it is part of the completion of this feature.

>     - This should take a dev week, if there are no major bugs or regressions.

Reviewer: my estimate is that this would take much, _much_ longer than this--the code change is easy, the implications are hard to manage and so is the testing.

# Decoupling VS builds of SDK projects
The behaviour of an SDK project, when built in Visual Studio and when built in the DotNet CLI can vary, as different pieces of imports can be fetched from either build agent. This situation is not great specially for some developers as it increases their coding workload. To solve this, we want to ensure that all logic from a build from an *SDK project* comes from the SDK, independently of where the build is processed.

## Goals and Motivation

We are aiming for:
 - Consistent end-user eperience for build in either DotNet CLI or Visual Studio.
 - Isolating Dotnet SDK build to only their components.
 - Decoupling the Dotnet SDK from VS.

> The reason we are persuing this is for a better experience when using or writting Roslyn analyzers and MSBuild Tasks. Currently tooling authors need to target NetStandard2.0 for their projects to be recognized by VS. Another options is to multi-target but it takes a lot more effort and time spent on that. Another aspect is the user experience, if the Roslyn version for VS analyzers and generators doesn't match the one in Visual Studio, they do not work.

Reviewer 2: This paragraph seems to be combining what TFM analyzers need to target and SDK / VS analyzer tearing. Think we should split up those concerns.

Reviewer: I agree. Roslyn analyzers/generators targeting netstandard2 are one concern, multitargeting tasks is another, SDK tearing is a third.

 - Reduced cost of development for external and internal teams that contribute to Roslyn Analyzers, SourceBuild generators, or MSBuild Tasks.
 - End-user will not experience mismatch between analyzer versions, and confirmation that their SDK style builds will behave the same way in VS and in the command line.

## Stakeholders
For the internal stakeholder, we have the teams that will continue the work to fully complete the VS and SDK decoupling feature after our base work is done. There are two handovers in this project:

1. Enabling the MSBuild.exe execution state to be the same as DotNet command line invocation so the Roslyn team can enable the use of their Core compiler in VS.
2. Tasks and other projects can be written in .NET core and the SDK projects will build successuflly in VS. This enables other teams to migrate their tasks to .NET core instead of keeping them targeting .NET Framework.

The handovers should allow other teams to proceed with their work smoothly and no change in build behaviour should be present within MSBuild.

## Risks
A few risks associated with this feature:

>  - There might be a performance hit on VS. It would depending on the amount of non-framwork tasks that the project will need to load when opening it in VS. The performance gain from pre-loading will not be available in this scenario.

Reviewer 2: This only manifests when we move tasks to .NET Core and use an MSBuild bridge for them correct? If so should we list out the phases / dials we have for this effort? The low end is just using .NTE Core tools and then we dial up to running tasks on core.

>  - The VS tooling might present have a different version than the SDK installed, which might lead to discrepancy in partial builds.

Reviewer 2: Part of our statement now is that VS and .NET SDK are de-coupled experiences so I wouldn't put this in the risk section. This is more leaning into our design of the system. I would put this in the impact section and say that an impcat of this change is it re-enforces our design that VS and .NET SDK be decoupled experiences.

Author: When you put it that way it sounds like a consequence of the design, which is not really covered in this doc specific. So I opted to take it out as a risk in this case.

 - Our work is early in the development effort. If this feature is discovered to have too large of an impact in experience of performance the work might be delayed or discarded.
 - There are no concrete deadlines for our part of the feature, but we are aiming for an early preview cycle, so we have a chance to measure the consequences and fix any issues that arise.

## Plan
 1. Ensure that MSBuild.exe provides the same execution state as the dotnet command line invocation.

>     -  This is should take around 1 dev week to complete, and will be handed over to Roslyn team.

Reviewer 2: Should we list explicitly `%DOTNET_BUILD_HOST%` here?

Reviewer 3: For a one-pager (high-level) I expected that it would be kept as a more general statement of "msbuild should align with the SDK's state set-up" and let details come out in a specific issue.
